py3status/modules/yubikey.py

```diff
@@ -0,0 +1,170 @@
+                    for event in inotify.read():
+                        # e.g. ssh doesn't start immediately, try several seconds to be sure
+                        for i in range(20):
+                            with this._check_card_status() as check:
+                                try:
+                                    # if this doesn't return very quickly, touch is pending
+                                    check.communicate(
+                                        timeout=self.gpg_check_timeout)
+                                    time.sleep(0.25)
+                                except subprocess.TimeoutExpired:
+                                    self.pending_gpg = True
+                                    os.killpg(check.pid, signal.SIGINT)
+                                    check.communicate()
+                                    # wait until device is touched
+                                    with this._check_card_status() as wait:
+                                        wait.communicate()
+                                        self.pending_gpg = False
+                                        break
+
+                        # ignore all events caused by operations above
+                        for _ in inotify.read():
+                            pass
+
+                for watch in watches:
+                    inotify.rm_watch(watch)
+
+        class SudoThread(threading.Thread):
+            def _restart_sudo_reset_timer(this):
+                def reset_pending_sudo():
+                    self.pending_sudo = False
+
+                if this.sudo_reset_timer is not None:
+                    this.sudo_reset_timer.cancel()
+
+                this.sudo_reset_timer = threading.Timer(5, reset_pending_sudo)
+                this.sudo_reset_timer.start()
+
+            def run(this):
+                this.sudo_reset_timer = None
+
+                inotify = inotify_simple.INotify()
+                watches = []
+                for path in self.sudo_check_watch_paths:
+                    if os.path.isfile(path):
+                        watch = inotify.add_watch(path,
+                                                  inotify_simple.flags.OPEN)
+                        watches.append(watch)
+
+                while not self.killed.is_set():
+                    for event in inotify.read():
+                        this._restart_sudo_reset_timer()
+                        self.pending_sudo = True
+
+                for watch in watches:
+                    inotify.rm_watch(watch)
+
+        if len(self.gpg_check_watch_paths) > 0:
+            GpgThread().start()
+
+        if len(self.sudo_check_watch_paths) > 0:
+            SudoThread().start()
+
+    def yubikey(self):
+        is_waiting = self.pending_gpg or self.pending_sudo
+        format_params = {'is_waiting': is_waiting}
+        response = {
+            'cached_until': self.py3.time_in(self.cache_timeout),
```

Comment:
@tobes, @lasers: I have to put a small number here, `1` or even smaller, but most of the time it would be wasting cycles and just redrawing the same output over and over again. I would really love to put here `self.py3.CACHE_FOREVER`, and instead just manually force module refresh when at least one checks is triggered in other threads - is this possible to achieve?

Reply:
Maybe you're looking for `self.py3.update()`? `scratchpad_async` and `window_title_async` might be interesting to you as both modules uses the function and the threading.

Reply:
Exactly, thanks! Now it is instantly reacting only when the value changes - awesome!!